File: Framework/L2DTargetPoint.cpp
/**
 *
 *  You can modify and use this source freely
 *  only for the development of application related Live2D.
 *
 *  (c) Live2D Inc. All rights reserved.
 */
#include "L2DTargetPoint.h"
#include "util/UtSystem.h"

namespace live2d
{
	namespace framework
	{
		L2DTargetPoint::L2DTargetPoint()
		{
			this->faceTargetX   = 0;
			this->faceTargetY   = 0;
			this->faceX         = 0;
			this->faceY         = 0;
			this->faceVX        = 0.0f;
			this->faceVY        = 0.0f;
			this->lastTimeSec   = 0;
		}
		
		
		
		void L2DTargetPoint::update()
		{
			
			
			const float FACE_PARAM_MAX_V = 40.0 / 10 ;
			const float MAX_V =  FACE_PARAM_MAX_V * 1.0f / FRAME_RATE ;
			
			static l2d_int64 lastTimeSec = 0 ;
			if( lastTimeSec == 0 )
			{
				lastTimeSec = UtSystem::getUserTimeMSec() ;
				return ;
			}
			l2d_int64 curTimeSec = UtSystem::getUserTimeMSec() ;
			
			float deltaTimeWeight = (float)(curTimeSec - lastTimeSec)*FRAME_RATE/1000.0f ;
			lastTimeSec = curTimeSec ;
			
			
			const float TIME_TO_MAX_SPEED = 0.15f ;
			const float FRAME_TO_MAX_SPEED = TIME_TO_MAX_SPEED * FRAME_RATE  ;//sec*frame/sec
			const float MAX_A = deltaTimeWeight * MAX_V / FRAME_TO_MAX_SPEED ;
			
			
			
			float dx = (faceTargetX - faceX) ;
			float dy = (faceTargetY - faceY) ;
			
			if( dx == 0 && dy == 0 ) return ;
			
			
			float d = L2D_SQRT( dx*dx + dy*dy ) ;
			
			
			float vx = MAX_V * dx / d ;
			float vy = MAX_V * dy / d ;
			
			
			float ax = vx - faceVX ;
			float ay = vy - faceVY ;
			
			float a = L2D_SQRT( ax*ax + ay*ay ) ;
			
			
			if( a < -MAX_A || a > MAX_A )
			{
				ax *= MAX_A / a ;
				ay *= MAX_A / a ;
			}
			
			
			faceVX += ax ;
			faceVY += ay ;
			
			
			
			
			
			{
				
				//            2  6           2               3
				//      sqrt(a  t  + 16 a h t  - 8 a h) - a t
				// v = --------------------------------------
				//                    2
				//                 4 t  - 2
				//(t=1)
				
				
				
				float max_v = 0.5f * ( L2D_SQRT( MAX_A*MAX_A + 16*MAX_A * d - 8*MAX_A * d ) - MAX_A ) ;
				float cur_v = L2D_SQRT( faceVX*faceVX + faceVY*faceVY ) ;
				
				if( cur_v > max_v )
				{
					
					faceVX *= max_v / cur_v ;
					faceVY *= max_v / cur_v ;
				}
			}
			
			faceX += faceVX ;
			faceY += faceVY ;
			
			return;
		}
		
		
		void L2DTargetPoint::set( float x, float y )
		{
			this->faceTargetX   = x;
			this->faceTargetY   = y;
		}
	}
}

File: package/live2d/v2/framework/motion/l2d_target_point.py
import math

from ...core import UtSystem


class L2DTargetPoint:
    FRAME_RATE = 30
    TIME_TO_MAX_SPEED = 0.15
    FACE_PARAM_MAX_V = 40.0 / 7.5
    MAX_V = FACE_PARAM_MAX_V / FRAME_RATE
    FRAME_TO_MAX_SPEED = TIME_TO_MAX_SPEED * FRAME_RATE

    def __init__(self):
        self.EPSILON = 0.01
        self.faceTargetX = 0
        self.faceTargetY = 0
        self.faceX = 0
        self.faceY = 0
        self.faceVX = 0
        self.faceVY = 0
        self.lastTimeSec = 0

    def setPoint(self, x, y):
        self.faceTargetX = x
        self.faceTargetY = y

    def getX(self):
        return self.faceX

    def getY(self):
        return self.faceY

    def update(self):

        if self.lastTimeSec == 0:
            self.lastTimeSec = UtSystem.getUserTimeMSec()
            return

        cur_time_sec = UtSystem.getUserTimeMSec()
        delta_time_weight = (cur_time_sec - self.lastTimeSec) * L2DTargetPoint.FRAME_RATE / 1000.0
        self.lastTimeSec = cur_time_sec
        max_a = delta_time_weight * self.MAX_V / self.FRAME_TO_MAX_SPEED
        dx = (self.faceTargetX - self.faceX)
        dy = (self.faceTargetY - self.faceY)
        if abs(dx) <= self.EPSILON and abs(dy) <= self.EPSILON:
            return
        d = math.sqrt(dx * dx + dy * dy)
        vx = self.MAX_V * dx / d
        vy = self.MAX_V * dy / d
        ax = vx - self.faceVX
        ay = vy - self.faceVY
        a = math.sqrt(ax * ax + ay * ay)
        if a < -max_a or a > max_a:
            ax *= max_a / a
            ay *= max_a / a

        self.faceVX += ax
        self.faceVY += ay

        max_v = 0.5 * (math.sqrt(max_a * max_a + 16 * max_a * d - 8 * max_a * d) - max_a)
        cur_v = math.sqrt(self.faceVX * self.faceVX + self.faceVY * self.faceVY)
        if cur_v > max_v:
            self.faceVX *= max_v / cur_v
            self.faceVY *= max_v / cur_v

        self.faceX += self.faceVX
        self.faceY += self.faceVY
